### PR TITLE
1033 validator question title 9

### PR DESCRIPTION
**Pull request by pricea1**
### What is the context of this PR?
Sets up foundation for validation of fields. Uses ajv JSON schema validation and Apollo client side cache. This PR only implements Page Title validation.

### How to review 
Check validation messages on page title match functionality of prototype: https://author.netlify.com

### CODE REVIEW

File: eq-author/src/App/QuestionnaireDesignPage/NavigationSidebar/PageNavItem.js [hunk 1]
```diff
@@ -25,10 +25,21 @@ const getIcon = pageType => {
   return PageIcon;
 };
 
+const getBadgeText = (page, canShowValidation) => {
```
Comment (tgandrews) [inline eq-author/src/App/QuestionnaireDesignPage/NavigationSidebar/PageNavItem.js]: This change is not required for this. You can remove it. We have a separate story that will probably use websockets to get this information for all future updates. So leave it off for the moment otherwise we are going to have to maintain it as we add more validations.

**Comment by heofs** [comment]
Metadata page seems to be broken.

File: eq-author-api/src/validation/schemas/entities.json [hunk 2]
```diff
@@ -0,0 +1,47 @@
+{
+  "$id": "http://example.com/schemas/entities.json",
+  "definitions": {
+    "answer": {
```
Comment (tgandrews) [inline eq-author-api/src/validation/schemas/entities.json]: We don't need to define anything but page title I would have thought. I probably did more in the spike just as I was playing around with it

File: eq-author/src/App/QuestionnaireDesignPage/NavigationSidebar/PageNavItem.js (hunk 1 above)
Comment (pricea1) [reply eq-author/src/App/QuestionnaireDesignPage/NavigationSidebar/PageNavItem.js]: Navigation badge now removed

File: eq-author-api/src/validation/schemas/entities.json (hunk 2 above)
Comment (pricea1) [reply eq-author-api/src/validation/schemas/entities.json]: I've removed answer schema def from this PR.

**Comment by pricea1** [comment]
Metadata page, Navigation link validation count and disabling of Launch button now removed from PR

File: eq-author/src/App/QuestionnaireDesignPage/NavigationSidebar/PageNavItem.js [hunk 3]
```diff
@@ -42,7 +42,9 @@ export const UnwrappedPageNavItem = ({
       icon={getIcon(page.pageType)}
       data-test="nav-page-link"
     >
-      {page.displayName}
+      {page.validationErrorInfo && page.validationErrorInfo.totalCount
+        ? `${page.displayName} (${page.validationErrorInfo.totalCount})`
```
[the remark this answers is not in the document]
Comment (pricea1) [reply eq-author/src/App/QuestionnaireDesignPage/NavigationSidebar/PageNavItem.js]: Now removed.

File: eq-author/src/components/withEntityEditor/index.js [hunk 4]
```diff
@@ -92,19 +110,30 @@ const withEntityEditor = entityPropName => WrappedComponent => {
...
 
       return (
-        <WrappedComponent
-          {...this.props}
-          {...props}
-          onChange={this.handleChange}
-          onUpdate={this.handleUpdate}
-          onSubmit={this.handleSubmit}
-        />
+        <div onBlur={this.handleLoseFocus}>
```
[the remark this answers is not in the document]
Comment (pricea1) [reply eq-author/src/components/withEntityEditor/index.js]: Fixed

**Comment by pricea1** [comment]
[Validator - Question title (9)](https://trello.com/c/v6ssj4E1/1033-validator-question-title-9)

File: eq-author-api/schema/resolvers/pages/questionPage.js [hunk 5]
```diff
@@ -89,12 +94,16 @@ Resolvers.Mutation = {
       typeof position === "number" ? position : section.pages.length;
     section.pages.splice(insertionPosition, 0, page);
     await saveQuestionnaire(ctx.questionnaire);
+
+    ctx.validationErrorInfo = await validateQuestionnaire(ctx.questionnaire);
```
Comment (tgandrews) [inline eq-author-api/schema/resolvers/pages/questionPage.js]: Happy to leave like this for this PR but we should probably look at moving the `saveQuestionnaire` and `validateQuestionnaire` logic to some point outside of the mutations.

File: eq-author-api/tests/utils/questionnaireBuilder/page/index.js [hunk 6]
```diff
@@ -15,4 +17,5 @@ module.exports = {
   duplicatePageMutation,
   movePage,
   movePageMutation,
+  validatePage,
```
Comment (tgandrews) [inline eq-author-api/tests/utils/questionnaireBuilder/page/index.js]: Does this need to be its own mutation? Can't we just read this information from the `queryPage`?

File: eq-author/src/App/QuestionnairesPage/withCreateQuestionnaire.js [hunk 7]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable graphql/template-strings */
```
Comment (tgandrews) [inline eq-author/src/App/QuestionnairesPage/withCreateQuestionnaire.js]: Is this needed?

File: eq-author/src/enhancers/withValidationError.test.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+import withValidationError from "./withValidationError";
+import VALIDATION_MESSAGES from "constants/validationMessages";
+
+const EnhancedComponent = withValidationError("div");
+
+const createWrapper = (props, render = shallow) =>
+  render(<EnhancedComponent {...props} />);
+
+describe("withValidationError", () => {
+  const props = {
```
Comment (tgandrews) [inline eq-author/src/enhancers/withValidationError.test.js]: Does this need to be such a complex object? Its not coupled to page AFAICT

File: eq-author/src/enhancers/withValidationError.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+import { propType } from "graphql-anywhere";
+import pageFragment from "graphql/fragments/page.graphql";
+
+import VALIDATION_MESSAGES from "constants/validationMessages";
+
+const withValidationError = WrappedComponent => {
+  return class extends React.Component {
+    static propTypes = {
+      enableValidationMessage: PropTypes.bool,
+      page: propType(pageFragment).isRequired,
```
Comment (tgandrews) [inline eq-author/src/enhancers/withValidationError.js]: Why is it only page?

Comment (tgandrews) [reply eq-author/src/enhancers/withValidationError.js]: Could we have it take any object and read its validationErrors? Or do you think thats too complex for the moment?

File: eq-author/src/App/QuestionnairesPage/withCreateQuestionnaire.js (hunk 7 above)
Comment (pricea1) [reply eq-author/src/App/QuestionnairesPage/withCreateQuestionnaire.js]: No, it was left over from trying different ways to fix the linting issue last week. Now removed.

File: eq-author-api/src/validation/schemas/entities.json [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+  "definitions": {
+    "page": {
+      "type": "object",
+      "properties": {
+        "id": { "type": "string" },
+        "title": {
+          "$ref": "definitions.json#/definitions/populatedString"
+        }
+      },
+      "required": ["id"]
```
Comment (tgandrews) [inline eq-author-api/src/validation/schemas/entities.json]: Is `title` also required?

File: eq-author/src/apollo/resolvers/index.js [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+      });
+
+      const updatedPageList = newPagesList.filter(
+        p => p !== _.updateQuestionPage.id
+      );
+
+      cache.writeData({ data: { newPagesList: updatedPageList } });
+    },
+
+    duplicatePage: (_, input, { cache }) => {
```
Comment (tgandrews) [inline eq-author/src/apollo/resolvers/index.js]: Do we need to do this? It could be fully completed.

File: eq-author/src/apollo/resolvers/index.js [hunk 12]
```diff
@@ -0,0 +1,125 @@
+import gql from "graphql-tag";
+
+export const typeDefs = gql`
+  extend type QuestionPage {
+    isNew: Boolean!
+  }
+
+  extend type Mutation {
```
Comment (tgandrews) [inline eq-author/src/apollo/resolvers/index.js]: Do we need to extend the mutations as well? I would assume we could just extend the type

File: eq-author/src/apollo/resolvers/index.js [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+  pages.forEach(questionPage => {
+    newPagesList.push(questionPage.id);
+  });
+
+  cache.writeData({ data: { newPagesList } });
+};
+
+const isNewQuestionPage = (questionPage, args, { cache }) => {
+  const { newPagesList, pagesVisitedList } = cache.readQuery({
```
Comment (tgandrews) [inline eq-author/src/apollo/resolvers/index.js]: I'm not sure I understand why we need to track the pages visited as well?

File: eq-author/src/components/withEntityEditor/index.js [hunk 14]
```diff
@@ -104,8 +128,19 @@ const withEntityEditor = entityPropName => WrappedComponent => {
...
+          enableValidationMessage={this.enableValidationMessage()}
         />
       );
+    };
+
+    render() {
+      const wrappedComponent = this.renderWrappedComponent();
+
+      return focusControl ? (
+        <div onBlur={this.handleLoseFocus}>{wrappedComponent}</div>
```
Comment (tgandrews) [inline eq-author/src/components/withEntityEditor/index.js]: We can't have `withEntityEditor` return DOM elements as we often use it to wrap table rows etc. It was originally intended to pass the functions down so the form could use them. Could we follow that pattern here?

@heofs I believe you may have looked into this around the metadata table

File: eq-author/src/enhancers/withValidationError.js (hunk 9 above)
Comment (pricea1) [reply eq-author/src/enhancers/withValidationError.js]: Changed to be generic object as per withEntityEditor

File: eq-author-api/src/validation/schemas/entities.json (hunk 10 above)
Comment (pricea1) [reply eq-author-api/src/validation/schemas/entities.json]: Added title to required

File: eq-author/src/apollo/resolvers/index.js (hunk 11 above)
Comment (pricea1) [reply eq-author/src/apollo/resolvers/index.js]: Removed duplicate functionality. Duplicated page is now not treated as new page

File: eq-author/src/apollo/resolvers/index.js (hunk 12 above)
Comment (pricea1) [reply eq-author/src/apollo/resolvers/index.js]: Apparently not. I've removed them

File: eq-author/src/apollo/resolvers/index.js (hunk 13 above)
Comment (pricea1) [reply eq-author/src/apollo/resolvers/index.js]: It was a workaround because new QuestionPage was returned twice - once for createQuestionPage mutation and again for subsequent getPage query. Refactored so that new page calls update mutation regardless of whether no values have changed. Now only need to track if page is new

File: eq-author/src/components/withEntityEditor/index.js (hunk 14 above)
Comment (pricea1) [reply eq-author/src/components/withEntityEditor/index.js]: After refactor where new page updates even when no change, we no longer need to track focus. I've removed this from withEntityEditor

File: eq-author-api/tests/utils/questionnaireBuilder/page/index.js (hunk 6 above)
Comment (pricea1) [reply eq-author-api/tests/utils/questionnaireBuilder/page/index.js]: I've modified queryPage test so it can be used for validationError test too. validatePage mutation now removed

File: eq-author/src/App/QuestionnaireDesignPage/index.js [hunk 15]
```diff
@@ -46,9 +46,13 @@ export class UnwrappedQuestionnaireDesignPage extends Component {
     }),
     location: PropTypes.object, // eslint-disable-line
     error: PropTypes.object, // eslint-disable-line
+    client: PropTypes.object, // eslint-disable-line
```
Comment (tgandrews) [inline eq-author/src/App/QuestionnaireDesignPage/index.js]: Why is this change needed?

File: eq-author/src/App/page/Design/QuestionPageEditor/MetaEditor.test.js [hunk 16]
```diff
@@ -16,6 +18,7 @@ describe("MetaEditor", () => {
       onUpdate: jest.fn(),
       fetchAnswers: jest.fn(),
       onChangeUpdate: jest.fn(),
+      getValidationError: getValidationError,
```
Comment (tgandrews) [inline eq-author/src/App/page/Design/QuestionPageEditor/MetaEditor.test.js]: Can we define it in beforeEach as well so that we don't have the call information shared between tests

Comment (tgandrews) [reply eq-author/src/App/page/Design/QuestionPageEditor/MetaEditor.test.js]: We also have no tests using function. Are we missing it?

File: eq-author/src/apollo/resolvers/index.js [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+  return newPagesList.indexOf(questionPage.id) > -1;
+};
+
+export const resolvers = {
+  QuestionPage: {
+    isNew: isNewQuestionPage,
+  },
+
+  Mutation: {
+    createQuestionnaire: (_, input, { cache }) => {
```
Comment (tgandrews) [inline eq-author/src/apollo/resolvers/index.js]: Normally `_` is used to denote an argument you are not using. What is it actually? Can we give it a name

File: eq-author/src/components/withEntityEditor/index.js [hunk 18]
```diff
@@ -92,7 +92,7 @@ const withEntityEditor = entityPropName => WrappedComponent => {
       this.props.onSubmit(this.entity);
     };
 
-    render() {
+    render = () => {
```
Comment (tgandrews) [inline eq-author/src/components/withEntityEditor/index.js]: This change isn't needed. Its not passed anywhere so does not need to be bound explicitly

File: eq-author/src/constants/validationMessages.js [hunk 19]
```diff
@@ -0,0 +1,3 @@
+export default {
+  ERR_VALID_REQUIRED: ({ label }) => `${label} is required`,
```
Comment (tgandrews) [inline eq-author/src/constants/validationMessages.js]: It would be nice to share this constant name between client and server. But I don't think we have a nice way of doing this yet.

File: eq-author/src/enhancers/withValidationError.js [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+    };
+
+    getValidationError = ({ field, label }) => {
+      const {
+        [entityPropName]: { validationErrorInfo },
+      } = this.props;
+
+      const messages = validationErrorInfo && validationErrorInfo.errors;
+
+      if (!messages || this.props[entityPropName].isNew) {
```
Comment (tgandrews) [inline eq-author/src/enhancers/withValidationError.js]: Minor nit: Could we extract `isNew` on line `19` as well?

**Comment by tgandrews** [comment]
Tested and works well in browser just final few small comments.

File: eq-author/src/components/withEntityEditor/index.js (hunk 18 above)
Comment (pricea1) [reply eq-author/src/components/withEntityEditor/index.js]: Accidentally left over from previous changes. Now reverted.

File: eq-author/src/apollo/resolvers/index.js (hunk 17 above)
Comment (pricea1) [reply eq-author/src/apollo/resolvers/index.js]: It was taken from Apollo docs I was following. Now changed to root variable name

File: eq-author/src/App/page/Design/QuestionPageEditor/MetaEditor.test.js (hunk 16 above)
Comment (pricea1) [reply eq-author/src/App/page/Design/QuestionPageEditor/MetaEditor.test.js]: Changed to jest.fn() as no tests using the function.

File: eq-author/src/App/QuestionnaireDesignPage/index.js (hunk 15 above)
Comment (pricea1) [reply eq-author/src/App/QuestionnaireDesignPage/index.js]: It isn't anymore. now removed.

File: eq-author/src/components/RichTextEditor/RichTextEditor.test.js [hunk 21]
```diff
@@ -549,4 +549,48 @@ describe("components/RichTextEditor", function() {
...
   });
+
+  describe("validation", () => {
+    it("should set input as valid if no validation message is passed as parameter", () => {
+      expect(wrapper.find("RichTextEditor__Input").prop("invalid")).toEqual(
+        false
+      );
+    });
+
+    it("should not display validation message is passed as parameter", () => {
```
Comment (heofs) [inline eq-author/src/components/RichTextEditor/RichTextEditor.test.js]: I dont get what this is testing. Could description be a little better?

File: eq-author-api/package.json [hunk 22]
```diff
@@ -13,6 +13,8 @@
     "create-migration": "node scripts/createMigration.js"
   },
   "dependencies": {
+    "ajv": "latest",
+    "ajv-errors": "^1.0.1",
```
Comment (samiwel) [inline eq-author-api/package.json]: Should this be set to "latest" too?

File: eq-author-api/schema/typeDefs.js [hunk 23]
```diff
@@ -43,6 +45,7 @@ type Questionnaire {
   shortTitle: String
   displayName: String!
   introduction: QuestionnaireIntroduction
+  isNew: Boolean
```
Comment (samiwel) [inline eq-author-api/schema/typeDefs.js]: This should be made mandatory i.e. `Boolean!`

File: eq-author-api/schema/typeDefs.js [hunk 24]
```diff
@@ -97,6 +100,8 @@ type QuestionPage implements Page {
   confirmation: QuestionConfirmation
   routing: Routing2
   totalValidation: TotalValidationRule
+  validationErrorInfo: ValidationErrorInfo
+  isNew: Boolean
```
Comment (samiwel) [inline eq-author-api/schema/typeDefs.js]: Same here, the `isNew` should be a mandatory boolean.

File: eq-author-api/schema/typeDefs.js [hunk 25]
```diff
@@ -202,6 +207,17 @@ type AvailableRoutingDestinations {
...
 
+type ValidationError {
+  field: String
+  errorCode: String
+}
+
+type ValidationErrorInfo {
+  errors: [ValidationError]
+  totalCount: Int
+}
```
Comment (samiwel) [inline eq-author-api/schema/typeDefs.js]:
```
type ValidationErrorInfo {
  errors: [ValidationError!]!
  totalCount: Int!
}
```

File: eq-author-api/schema/typeDefs.js [hunk 26]
```diff
@@ -202,6 +207,17 @@ type AvailableRoutingDestinations {
   sections: [Section]!
 }
 
+type ValidationError {
+  field: String
+  errorCode: String
+}
```
Comment (samiwel) [inline eq-author-api/schema/typeDefs.js]:
```
type ValidationError {
  field: String!
  errorCode: String!
}
```

Comment (pricea1) [reply eq-author-api/schema/typeDefs.js]: Done

File: eq-author-api/schema/typeDefs.js (hunk 25 above)
Comment (pricea1) [reply eq-author-api/schema/typeDefs.js]: Done

File: eq-author-api/schema/typeDefs.js (hunk 24 above)
Comment (pricea1) [reply eq-author-api/schema/typeDefs.js]: done

File: eq-author-api/schema/typeDefs.js (hunk 23 above)
Comment (pricea1) [reply eq-author-api/schema/typeDefs.js]: done

File: eq-author-api/package.json (hunk 22 above)
Comment (pricea1) [reply eq-author-api/package.json]: done

File: eq-author/src/components/RichTextEditor/RichTextEditor.test.js (hunk 21 above)
Comment (pricea1) [reply eq-author/src/components/RichTextEditor/RichTextEditor.test.js]: I've changed it to be more descriptive.

File: eq-author/src/App/page/Design/QuestionPageEditor/AnswersEditor/index.js [hunk 27]
```diff
@@ -18,7 +19,7 @@ export const AnswersEditor = ({
   onDeleteAnswer,
   moveAnswer,
 }) => {
-  return (
+  return isEmpty(answers.length) ? (
```
Comment (samiwel) [inline eq-author/src/App/page/Design/QuestionPageEditor/AnswersEditor/index.js]: Should this be `isEmpty(answers)`?

Comment (samiwel) [reply eq-author/src/App/page/Design/QuestionPageEditor/AnswersEditor/index.js]: Actually, should the logic be reversed? `!isEmpty(answers)`

File: eq-author/src/enhancers/withValidationError.js (hunk 20 above)
Comment (pricea1) [reply eq-author/src/enhancers/withValidationError.js]: done

File: eq-author/src/App/page/Design/QuestionPageEditor/AnswersEditor/index.js (hunk 27 above)
Comment (pricea1) [reply eq-author/src/App/page/Design/QuestionPageEditor/AnswersEditor/index.js]: now reverted. don't think its needed anymore.
